File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:

void solve(vector<vector<char>>& grid,int i,int j,int r,int c){
    queue<pair<int,int>>q;
    q.push({i,j});

    while(!q.empty()){
        auto t=q.front(); q.pop();
        int x=t.first;
        int y=t.second;
        // grid[x][y]='2';

        if(y+1<c && grid[x][y+1]=='1'  ){
            q.push({x,y+1});         grid[x][y+1]='2';

        }
        if(y>0 && grid[x][y-1]=='1'){
            q.push({x,y-1});         grid[x][y-1]='2';

        }
        if(x+1<r && grid[x+1][y]=='1'){
            q.push({x+1,y});         grid[x+1][y]='2';

        }
        if(x>0 && grid[x-1][y]=='1' ){
            q.push({x-1,y});         grid[x-1][y]='2';

        }

    }
}
    int numIslands(vector<vector<char>>& grid) {

        int r=grid.size();
        int c=grid[0].size();


        int res=0;

        for(int i=0;i<r;i++){
            for(int j=0;j<c;j++){
                if(grid[i][j]=='1'){
                    solve(grid,i,j,r,c);
                    res++;
                }
            }
        }
        return res;
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    boolean check(char[][]grid,boolean [][] vis ,int x,int y){
        int n=grid.length;
        int m=grid[0].length;

        if(x<0 || x>=n || y<0 || y>=m || grid[x][y]=='0'|| vis[x][y]==true)return false;
        return true;
    }
    void solve(char[][]grid,boolean [][] vis ,int i,int j){

        vis[i][j]=true;

        int []fir={1,-1,0,0};
        int []sec={0,0,1,-1};

        for(int a=0;a<4;a++){
            int x=i+fir[a];
            int y=j+sec[a];

            if(check(grid,vis,x,y)){
                solve(grid,vis,x,y);
            }
        }

    }
    public int numIslands(char[][] grid) {
        int n=grid.length;
        int m=grid[0].length;
        boolean [][]vis=new boolean[n][m];

        int c=0;

        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(vis[i][j]!=true && grid[i][j]=='1'){
                    solve(grid,vis,i,j);
                    c++;
                }
            }
        }
        return c;
        
    }
}